### PR TITLE
Exposing rest-area names in passive maneuvers

### PR DESCRIPTION
**Pull request by ktatterso**
For various reasons I closed my previous PR and reopened this new one.  The changes are effectively the same. 

# Issue

We currently support returning `{"service", "rest_area"}` for in our routing response.  This PR adds support for also returning the name of the rest-area (if one exists).

The gist of the changes:
- When building "intersecting edges" in trip-leg-builder, add sign information if requested. (The rest-area name is on the sign.)
- To accomplish the previous step required adding a `Sign` to the `TripLeg_IntersectingEdge` protobuf object.  (Which is then passed from thor to odin.)
- When creating/serializing the routing response (in odin, route_serializer_osrm.cc) the passive maneuver for the rest-area, add the rest-area name the the response.

I sort-of apologize in advance for some refactoring that makes this look like a lot of changes.  I simply felt that some of the routines in triplegbuilder_utils.h should instead live in triplegbuilder.cc.

I will add comments to the diffs that will help steer the reviewer(s) to the code changes of interest. 

# Testing

This was straightforward.  There were existing gurka tests with a rest-area name.  All I did was add an assertion that we returned the rest-area name.

RAD/postman tests passed.

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/thor/triplegbuilder.cc [hunk 1]
```diff
@@ -462,6 +462,285 @@ void SetHeadings(TripLeg_Edge* trip_edge,
   }
 }
 
+// Walk the edge_signs, add sign information onto the trip_sign, honoring which to
+// add per the attributes-controller.
+void AddSignInfo(const AttributesController& controller,
```
Comment (ktatterso) [inline src/thor/triplegbuilder.cc:467]: Broke this out into its own routine since it is called from two places.

File: src/thor/triplegbuilder.cc [hunk 2]
```diff
@@ -462,6 +462,285 @@ void SetHeadings(TripLeg_Edge* trip_edge,
...
+ * @param  controller   Controller to determine which attributes to set.
+ * @param  directededge Directed edge on the path.
+ * @param  prev_de  Previous directed edge on the path.
+ * @param  local_edge_index  Index of the local intersecting path edge at intersection.
+ * @param  nodeinfo  Node information of the intersection.
+ * @param  trip_node  Trip node that will store the intersecting edge information.
+ * @param  intersecting_de Intersecting directed edge. Will be nullptr except when
+ *                         on the local hierarchy.
+ */
+void AddTripIntersectingEdge(const AttributesController& controller,
```
Comment (ktatterso) [inline src/thor/triplegbuilder.cc:558]: Moved from triplegbuilder_utils.h.

File: src/thor/triplegbuilder.cc [hunk 3]
```diff
@@ -462,6 +462,285 @@ void SetHeadings(TripLeg_Edge* trip_edge,
...
+ * @param controller               tells us what info we should add about the intersecting edges
+ * @param start_tile               the tile which contains the node
+ * @param node                     the node at which we are copying intersecting edges
+ * @param directededge             the current edge leaving the current node in the path
+ * @param prev_de                  the previous edge in the path
+ * @param prior_opp_local_index    opposing edge local index of previous edge in the path
+ * @param graphreader              graph reader for graph access
+ * @param trip_node                pbf node in the pbf structure we are building
+ */
+void AddIntersectingEdges(const AttributesController& controller,
```
Comment (ktatterso) [inline src/thor/triplegbuilder.cc:664]: Moved from triplegbuilder_utils.h.

File: src/thor/triplegbuilder.cc [hunk 4]
```diff
@@ -462,6 +462,285 @@ void SetHeadings(TripLeg_Edge* trip_edge,
...
+  if (controller.attributes.at(kNodeIntersectingEdgeRoadClass)) {
+    intersecting_edge->set_road_class(GetRoadClass(intersecting_de->classification()));
+  }
+
+  // Set the lane count for the intersecting edge if requested
+  if (controller.attributes.at(kNodeIntersectingEdgeLaneCount)) {
+    intersecting_edge->set_lane_count(intersecting_de->lanecount());
+  }
+
+  // Set the sign info for the intersecting edge if requested
```
Comment (ktatterso) [inline src/thor/triplegbuilder.cc:637]: Here's some new logic.  This gets the signs from the `intersecting_de` and puts them on the `valhalla::TripSign`.

File: src/thor/triplegbuilder.cc [hunk 5]
```diff
@@ -536,80 +815,8 @@ TripLeg_Edge* AddTripEdge(const AttributesController& controller,
     // Add the edge signs
     std::vector<SignInfo> edge_signs = graphtile->GetSigns(idx);
     if (!edge_signs.empty()) {
-      valhalla::TripSign* trip_sign = trip_edge->mutable_sign();
```
Comment (ktatterso) [inline src/thor/triplegbuilder.cc:539]: Moved to triplegbuilder.cc.

File: src/thor/triplegbuilder_utils.h [hunk 6]
```diff
@@ -334,181 +334,4 @@ struct MultimodalBuilder {
   }
 };
 
-/**
```
Comment (ktatterso) [inline src/thor/triplegbuilder_utils.h:337]: Moved to triplegbuilder.cc.

File: src/thor/triplegbuilder_utils.h [hunk 7]
```diff
@@ -334,181 +334,4 @@ struct MultimodalBuilder {
...
-    itersecting_edge->set_road_class(GetRoadClass(intersecting_de->classification()));
-  }
-
-  // Set the lane count for the intersecting edge if requested
-  if (controller.attributes.at(kNodeIntersectingEdgeLaneCount)) {
-    itersecting_edge->set_lane_count(intersecting_de->lanecount());
-  }
-}
-
-/**
```
Comment (ktatterso) [inline src/thor/triplegbuilder_utils.h:426]: Moved to triplegbuilder.cc.

File: src/tyr/route_serializer_osrm.cc [hunk 8]
```diff
@@ -546,6 +548,15 @@ json::ArrayPtr intersections(const valhalla::DirectionsLeg::Maneuver& maneuver,
 
         if (routeable && intersecting_edge->use() == TripLeg_Use_kRestAreaUse) {
           rest_stop->emplace("type", std::string("rest_area"));
+          if (intersecting_edge->has_sign()) {
```
Comment (ktatterso) [inline src/tyr/route_serializer_osrm.cc:551]: Here's where we check if the `intersecting_edge` has a sign, and if so, grab the name of the sign and add it to the response.

File: test/gurka/test_use.cc [hunk 9]
```diff
@@ -90,6 +90,7 @@ TEST_F(Use, test_passing_rest_area) {
   EXPECT_FALSE(steps[step_idx]["intersections"][0].HasMember("rest_stop"));
   EXPECT_TRUE(steps[step_idx]["intersections"][1].HasMember("rest_stop"));
   EXPECT_STREQ(steps[step_idx]["intersections"][1]["rest_stop"]["type"].GetString(), "rest_area");
+  EXPECT_STREQ(steps[step_idx]["intersections"][1]["rest_stop"]["name"].GetString(), "Bear Peak");
```
Comment (ktatterso) [inline test/gurka/test_use.cc:93]: The line of code that proves we can retrieve the rest-area name.

File: proto/trip.proto [hunk 10]
```diff
@@ -250,6 +250,7 @@ message TripLeg {
     optional Use use = 7;
     optional RoadClass road_class = 8;
     optional uint32 lane_count = 9;
+    optional TripSign sign = 10;
```
Comment (ktatterso) [inline proto/trip.proto:253]: Now the `IntersectingEdge` can store sign information.

**Review by mandeepsandhu — APPROVED** [review]
lgtm

**Pull request merged**